crates/ra_editor/src/lib.rs

```diff
@@ -107,21 +107,10 @@ pub fn highlight(file: &File) -> Vec<HighlightedRange> {
 }
 
 pub fn diagnostics(file: &File) -> Vec<Diagnostic> {
-    let mut res = Vec::new();
-
-    for node in file.syntax().descendants() {
-        if node.kind() == ERROR {
-            res.push(Diagnostic {
-                range: node.range(),
-                msg: "Syntax Error".to_string(),
-            });
-        }
-    }
-    res.extend(file.errors().into_iter().map(|err| Diagnostic {
+    file.errors().into_iter().map(|err| Diagnostic {
         range: TextRange::offset_len(err.offset, 1.into()),
-        msg: err.msg,
-    }));
-    res
+        msg: "Syntax Error: ".to_string() + &err.msg,
+    }).collect()
```

Comment:
I assume that `file.errors()` has the same elements as `file.syntax().descendants().filter(|n| n.kind() == ERROR)`. If this is not true, I think we should implement an error kind enum to differentiate between syntax errors and other errors.

Reply:
They are slightly different things. `n.kind() == ERROR` are explicit error nodes used when we skip tokens. Roughly, in

```
some garbadge

fn foo() {}
```

`some garbadge` will be inside an explicit `ERROR` node. 


`file.errors()` is a list of error messages. I think we have a message for each `ERROR` node, but there are also messages which don't have the corresponding `ERROR` node, when some tokens are missing:

![image](https://user-images.githubusercontent.com/1711539/46794036-a9391100-cd4f-11e8-80d9-08e30f116fa7.png)